### ISSUE TITLE
Warn for nonstandard __exit__ argument names

### ISSUE THREAD

**Issue by jamesbraza** [PyCQA/pylint#7534]
### Current problem

I have seen many variations of argument names in `__exit__` signatures:

```python
class ContextManaged:

    def __exit__(exc_type, exc_value, traceback):
        ...  # SEE: https://docs.python.org/3/reference/datamodel.html#object.__exit__

    def __exit__(self, exception_type, exception_value, exception_traceback):
        ...  # SEE: https://www.geeksforgeeks.org/__exit__-in-python/

    def __exit__(self, t, value, traceback):
        ...  # SEE: https://github.com/python/mypy/issues/4885

    def __exit__(self, exc_type, exc_val, exc_tb):
        ...  # PyCharm autocomplete

    def __exit__(self, t, value, tb):
        ...
```

IMO there should be standardization here, as `__exit__` overrides a method from `object`.  I think pylint can help enforce this.

### Desired solution

I think pylint should emit a refactor or warning message for any argument names in `__exit__` signatures that don't match the [Python data model `__exit__` section](https://docs.python.org/3/reference/datamodel.html#object.__exit__).

```python
    def __exit__(exc_type, exc_value, traceback):
        ...
```

I am not sure which refactor message to emit (maybe make a new one), or alternately the [`arguments-renamed`](https://pylint.pycqa.org/en/latest/user_guide/messages/warning/arguments-renamed.html) warning message could work too.

**Comment by DanielNoord** [comment]
If we want to get really fancy we could dop something with https://github.com/PyCQA/astroid/pull/1810 and create a `conforms_to_model` helper function, but I don't think we should.

Agree that this would be a good warning and I think `arguments-renamed` makes sense. We would need to update the documentation about that message slightly though.

**Comment by jamesbraza** [comment]
Thanks for the response, I like all the points you make!  Here's the current message for `arguments-renamed`:

> Used when a method parameter has a different name than in the implemented interface or in an overridden method.

It seems to be pretty good to me, I am just curious, what makes you think it needs updating?  I see `object.__exit__` as just another instance of mismatching name(s) being used in an overridden method.

**Comment by DanielNoord** [comment]
Yeah, the issue is that `object.__exit__` doesn't exist 😅
```console
❯ python
Python 3.11.0b4 (main, Jul 12 2022, 11:43:04) [Clang 13.1.6 (clang-1316.0.21.2.5)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> object.__exit__
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: type object 'object' has no attribute '__exit__'. Did you mean: '__init__'?
```

It comes from something else and is added when creating a context manager. So, technically, `ContextManaged.__exit__` is not overwriting any interface, which is also why it isn't currently flagged. There is nothing in the `mro` being overwritten.

I do still think we should consider it as such, but we just need to special case it in the checker.

**Comment by jamesbraza** [comment]
Ha that's funny, considering the [Python data model page for `__exit__`](https://docs.python.org/3/reference/datamodel.html#object.__exit__) literally says:

> `object.__exit__(self, exc_type, exc_value, traceback)`

I agree with what all you're saying, and it sounds good to me.  Sorry that I am not helping, consider myself just a reporter here.  Thanks again for all your work @DanielNoord !

**Comment by DanielNoord** [comment]
No worries! In the spirit of reporting you might also want to check if there is an issue about this in the cpython repository. They might want to update their documentation to make this a bit more explicit. If you do please tag me or link it here as their response might help us with our own object model implementation.

**Comment by jamesbraza** [comment]
Yes I surfed through their issues for a while, and couldn't find one talking about this, so I went ahead and made https://github.com/python/cpython/issues/97699

A fun little rabbit hole!
